File: Languages/python/algorithms-dsa/leetcode/dynamic-programming/climbing_stairs.py
"""
LeetCode #70 - Climbing Stairs
Topic: Dynamic Programming
Difficulty: Easy

Count ways to climb n stairs (1 or 2 steps at a time) using fibonacci DP.

Time Complexity: O(n)
Space Complexity: O(1)
"""


def climb_stairs(n: int) -> int:
    if n <= 2:
        return n
    prev, curr = 1, 2
    for _ in range(3, n + 1):
        prev, curr = curr, prev + curr
    return curr


if __name__ == "__main__":
    assert climb_stairs(2) == 2
    assert climb_stairs(3) == 3
    assert climb_stairs(5) == 8
    assert climb_stairs(1) == 1
    print("All tests passed!")

File: Languages/cpp/algorithms-dsa/leetcode/dynamic-programming/climbing_stairs.cpp
/**
 * LeetCode 70: Climbing Stairs
 * Topic: Dynamic Programming
 * Difficulty: Easy
 *
 * You can climb 1 or 2 steps. How many distinct ways to reach the top?
 * Time: O(n), Space: O(1)
 */
#include <iostream>
#include <cassert>
using namespace std;

int climbStairs(int n) {
    if (n <= 2) return n;
    int prev2 = 1, prev1 = 2;
    for (int i = 3; i <= n; i++) {
        int curr = prev1 + prev2;
        prev2 = prev1;
        prev1 = curr;
    }
    return prev1;
}

int main() {
    assert(climbStairs(1) == 1);
    assert(climbStairs(2) == 2);
    assert(climbStairs(3) == 3);
    assert(climbStairs(5) == 8);
    assert(climbStairs(10) == 89);

    cout << "All tests passed!" << endl;
    return 0;
}
